Plutarch/TryFrom.hs

```diff
@@ -1,368 +1,77 @@
-{-# LANGUAGE AllowAmbiguousTypes #-}
-{-# LANGUAGE FunctionalDependencies #-}
 {-# LANGUAGE UndecidableInstances #-}
 
 module Plutarch.TryFrom (
   PTryFrom (..),
-  HRecP (..),
   ptryFrom,
+  PSubtype,
+  pupcast,
+  pupcastF,
 ) where
 
-import Data.Proxy (Proxy (Proxy))
-
-import GHC.TypeLits (KnownNat, Nat, Symbol, natVal, type (+))
-
-import Plutarch.Unsafe (punsafeCoerce, punsafeFrom)
-
-import Plutarch.Bool (pif, (#==))
-import Plutarch.ByteString (PByteString)
-import Plutarch.Integer (PInteger)
-import Plutarch.Unit (PUnit (PUnit))
+import Data.Coerce (Coercible)
+import Data.Kind (Constraint)
+import Data.Proxy (Proxy)
 
+import Plutarch.Internal (punsafeCoerce)
 import Plutarch.Internal.Other (
   DerivePNewtype,
   PInner,
   POpaque,
   PType,
-  S,
   Term,
-  pcon,
-  pdelay,
-  pforce,
-  plam,
-  plet,
-  popaque,
-  (#),
-  type (:-->),
  )
 
-import Plutarch.Trace (ptraceError)
-
-import Plutarch.DataRepr.Internal.HList (HRec (HCons, HNil), HRecGeneric (HRecGeneric), Labeled (Labeled))
-
-import Plutarch.DataRepr.Internal (
-  PDataRecord,
-  PDataSum,
-  PLabeledType ((:=)),
-  pdnil,
- )
+import Plutarch.Reducible (Reducible (Reduce))
 
-import Plutarch.List (
-  phead,
-  pmap,
-  ptail,
- )
+class PSubtypeDecl (a :: PType) (b :: PType)
 
-import Plutarch.Builtin (
-  PAsData,
-  PBuiltinList,
-  PBuiltinMap,
-  PBuiltinPair,
-  PData,
-  PIsData (pfromData),
-  pasByteStr,
-  pasConstr,
-  pasInt,
-  pasList,
-  pasMap,
-  pchooseListBuiltin,
-  pdata,
-  pforgetData,
-  pfstBuiltin,
-  ppairDataBuiltin,
-  psndBuiltin,
- )
+-- Not `OVERLAPPABLE` or `OVERLAPPING` on purpose
+instance {-# OVERLAPS #-} PSubtypeDecl a a
 
-import Plutarch.TermCont (TermCont (TermCont, runTermCont), tcont, unTermCont)
+-- FIXME: Relax subtyping constraint to a @c@ rather than all @c@.
+-- This is currently not really possible because a type must have
+-- exactly one super type and no more.
+instance {-# OVERLAPS #-} (PSubtypeDecl a b, PInner c POpaque ~ b) => PSubtypeDecl a c
 
-import Plutarch.DataRepr.Internal (PIsDataRepr (PIsDataReprRepr), PIsDataReprInstances)
-
-import Plutarch.Reducible (Reducible (Reduce))
+{- | @PSubtype a b@ constitutes a subtyping relation between @a@ and @b@.
+ This concretely means that `\(x :: Term s b) -> punsafeCoerce x :: Term s a`
+ is legal and sound.
 
-import Data.Functor.Const (Const)
+ You can not make an instance for this yourself.
+ You must use the 'PInner' type family of 'PlutusType' to get this instance.
 
-import Data.Kind (Type)
+ Caveat: Only @PInner a POpaque@ is considered unfortunately, as otherwise
+ getting GHC to figure out the relation with multiple supertypes is quite hard.
```

Comment:
The diamond problem? I feel like we shouldn't allow multiple supertypes but then again a fundep would be too restrictive as well. I think Arc was actually talking about this exact issue and he managed to solve it with the dual typeclasses AFAIK - but that didn't work in this case without putting GHC into a loop?

I'm concerned about the type inference and potential complexity.

Reply:
I've decided to drop multiple supertypes entirely for the sake of #82